### PR TITLE
feat: add data availability

### PR DESCRIPTION
**Pull request by noel2004**
BREAKING: the new feature require more input field in witness so the prover-cluster would be broken for this new circuit.
Merging should be suspended until rollup manager has supported data availability (https://github.com/fluidex/rollup-state-manager/issues/178)

This work is mainly for issue #144, include following parts:

- [X] Improve decode_tx circuit to encode tx data which would be put on chain

- [x] add hash_input circuit to hash tx data

- [x] put things together into block circuit, make hash an public output signal

(All of works above should include test cases)

- [x] tools which encoding tx data for L1 tx and verifications

For first phase we just put transfer on chain, and data encoded as follows:

> [ NAccountLevels  bits ] fromAccountID
> [ NAccountLevels  bits ] toAccountID
> [ 16 bits ] tokenID
> [ 40 bits ] amount

We have no token records on L2 so just purpose the maximum count is 2^16 (2 bytes, according to the discussion in issue)

### CODE REVIEW

**Comment by noel2004** [comment]
When data avaliability has been induced, about 30k constraints has been added into the circuit for block 2_2_2_2

should notice that the size should increase linearly with nTxs

> [INFO]  snarkJS: Curve: bn-128
> [INFO]  snarkJS: # of Wires: 81318
> [INFO]  snarkJS: # of Constraints: 81804
> [INFO]  snarkJS: # of Private Inputs: 164
> [INFO]  snarkJS: # of Public Inputs: 2
> [INFO]  snarkJS: # of Labels: 447045
> [INFO]  snarkJS: # of Outputs: 2

File: tpl/ejs/src/decode_tx.circom.ejs [hunk 1]
```diff
@@ -1,19 +1,51 @@
...
 /**
  * @input in - {Array(Field)} - encoded transaction
  */
 
 function TxLength() { return <%= codegen.config.commonPayload.length %>; }
+<%= codegen.codeBlocks.DAProtocolUtils %>
+
+template DecodeTx(balanceLevels, orderLevels, accountLevels) {
+    //TODO: we should also have levels for token tree?
+    var tokenLevels = 16;
```
Comment (HAOYUatHZ) [inline tpl/ejs/src/decode_tx.circom.ejs:14]: `tokenLevels` is `balanceLevels`

File: test/testcases/hash_bytes.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+import * as path from 'path';
```
Comment (HAOYUatHZ) [inline test/testcases/hash_bytes.ts:1]: this test is not run, and seems incorrect? (correct version is in test/testcases/hash_txdata.ts).

seems we should remove this file?

File: test/codec/float.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  let m = f;
+  let e = 0;
+
+  while (Scalar.isZero(Scalar.mod(m, 10)) && !Scalar.isZero(Scalar.div(m, 0x800000000))) {
+    m = Scalar.div(m, 10);
+    e++;
+  }
+
+  if (e > 31) {
```
Comment (HAOYUatHZ) [inline test/codec/float.ts:47]: `if (e > mantisaLen)` ?

File: test/codec/float.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+  let m = f;
+  let e = 0;
+
+  while (!Scalar.isZero(Scalar.div(m, mantisa))) {
+    m = Scalar.div(m, 10);
+    e++;
+  }
+
+  if (e > 31) {
```
Comment (HAOYUatHZ) [inline test/codec/float.ts:76]: `if (e > mantisaLen)` ?

File: test/codec/float.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+  let e = 0;
+
+  while (!Scalar.isZero(Scalar.div(m, mantisa))) {
+    const roundUp = Scalar.gt(Scalar.mod(m, 10), 5);
+    m = Scalar.div(m, 10);
+    if (roundUp) m = Scalar.add(m, 1);
+    e++;
+  }
+
+  if (e > 31) {
```
Comment (HAOYUatHZ) [inline test/codec/float.ts:103]: `if (e > mantisaLen)` ?

**Comment by noel2004** [comment]
Has made fixing according to the comments

The missed decode circuit for float40 is completed now, but many templatize works on circuits still being left (now they are mostly hard-coded) ...

**Review by HAOYUatHZ — APPROVED** [review]

**Review by lispc — APPROVED** [review]

**Pull request merged**